File: internal/services/bigquery/job_handlers.go
package bigquery

import (
	"bytes"
	"errors"
	"fmt"
	"net/http"
	"net/url"
	"os"
	"strconv"
	"strings"
	"time"

	"devcloud/internal/events"
)

func (s *Server) insertJob(w http.ResponseWriter, r *http.Request, projectID string) {
	request, err := s.decodeJobInsertRequest(r.Body)
	if err != nil {
		writeError(w, http.StatusBadRequest, "invalid", "invalid json request")
		return
	}
	if request.JobReference.ProjectID != "" && request.JobReference.ProjectID != projectID {
		writeError(w, http.StatusBadRequest, "invalid", "jobReference.projectId must match request project")
		return
	}
	switch {
	case strings.TrimSpace(request.Configuration.Query.Query) != "":
		job, err := s.createQueryJob(projectID, request.JobReference, request.Configuration.Query, 0, true, request.Configuration.DryRun, s.effectiveUseLegacySQL(request.Configuration.Query.UseLegacySQL))
		if err != nil {
			writeError(w, http.StatusBadRequest, "invalidQuery", err.Error())
			return
		}
		events.Emit(s.eventPublisher, events.Event{
			Type:    "bigquery.job.inserted",
			Service: "bigquery",
			Payload: map[string]any{"project": projectID, "jobType": "query"},
		})
		writeJSON(w, http.StatusOK, job.Job)
	case request.Configuration.Copy.DestinationTable.TableID != "":
		job, err := s.createCopyJob(projectID, request.JobReference, request.Configuration.Copy)
		if err != nil {
			writeError(w, http.StatusBadRequest, "invalid", err.Error())
			return
		}
		events.Emit(s.eventPublisher, events.Event{
			Type:    "bigquery.job.inserted",
			Service: "bigquery",
			Payload: map[string]any{"project": projectID, "jobType": "copy"},
		})
		writeJSON(w, http.StatusOK, job.Job)
	case request.Configuration.Load.DestinationTable.TableID != "":
		job, err := s.createLoadJob(r.Context(), projectID, request.JobReference, request.Configuration.Load)
		if err != nil {
			writeError(w, http.StatusBadRequest, "invalid", err.Error())
			return
		}
		events.Emit(s.eventPublisher, events.Event{
			Type:    "bigquery.job.inserted",
			Service: "bigquery",
			Payload: map[string]any{"project": projectID, "jobType": "load"},
		})
		writeJSON(w, http.StatusOK, job.Job)
	case request.Configuration.Extract.SourceTable.TableID != "":
		job, err := s.createExtractJob(r.Context(), projectID, request.JobReference, request.Configuration.Extract)
		if err != nil {
			writeError(w, http.StatusBadRequest, "invalid", err.Error())
			return
		}
		events.Emit(s.eventPublisher, events.Event{
			Type:    "bigquery.job.inserted",
			Service: "bigquery",
			Payload: map[string]any{"project": projectID, "jobType": "extract"},
		})
		writeJSON(w, http.StatusOK, job.Job)
	default:
		writeError(w, http.StatusBadRequest, "invalid", "configuration.query.query, configuration.copy, configuration.load, or configuration.extract is required")
	}
}

func (s *Server) insertUploadJob(w http.ResponseWriter, r *http.Request, projectID string) {
	if r.URL.Query().Get("uploadType") != "multipart" {
		writeError(w, http.StatusBadRequest, "invalid", "uploadType=multipart is required")
		return
	}
	request, media, err := s.decodeMultipartJobInsertRequest(w, r)
	if err != nil {
		writeError(w, http.StatusBadRequest, "invalid", "invalid multipart upload request")
		return
	}
	if request.JobReference.ProjectID != "" && request.JobReference.ProjectID != projectID {
		writeError(w, http.StatusBadRequest, "invalid", "jobReference.projectId must match request project")
		return
	}
	if request.Configuration.Load.DestinationTable.TableID == "" {
		writeError(w, http.StatusBadRequest, "invalid", "configuration.load.destinationTable is required")
		return
	}
	job, err := s.createUploadLoadJob(projectID, request.JobReference, request.Configuration.Load, bytes.NewReader(media))
	if err != nil {
		writeError(w, http.StatusBadRequest, "invalid", err.Error())
		return
	}
	writeJSON(w, http.StatusOK, job.Job)
}

func (s *Server) createCompletedJob(requestProjectID string, requestedRef jobReference, config jobConfiguration, statistics jobStatistics) (queryJobRecord, error) {
	now := time.Now().UTC()
	jobID := strings.TrimSpace(requestedRef.JobID)
	if jobID == "" {
		jobID = "devcloud_job_" + strconv.FormatInt(now.UnixNano(), 10)
	} else if err := validateResourceID(jobID, "job"); err != nil {
		return queryJobRecord{}, err
	} else if _, found, err := s.readQueryJob(requestProjectID, jobID); err != nil {
		return queryJobRecord{}, err
	} else if found {
		return queryJobRecord{}, fmt.Errorf("already exists: job %s:%s", requestProjectID, jobID)
	}
	jobRef := jobReference{
		ProjectID: requestProjectID,
		JobID:     jobID,
		Location:  defaultString(requestedRef.Location, s.defaultLocation()),
	}
	resource := jobResource{
		Kind:          "bigquery#job",
		ID:            requestProjectID + ":" + jobID,
		SelfLink:      "/bigquery/v2/projects/" + url.PathEscape(requestProjectID) + "/jobs/" + url.PathEscape(jobID),
		JobReference:  jobRef,
		Configuration: config,
		Status:        jobStatus{State: "DONE"},
		Statistics:    statistics,
	}
	job := queryJobRecord{
		Job: resource,
		Response: queryResponse{
			Kind:         "bigquery#queryResponse",
			JobReference: jobRef,
			TotalRows:    "0",
			JobComplete:  true,
			CacheHit:     false,
		},
	}
	if err := s.writeQueryJob(requestProjectID, jobID, job); err != nil {
		return queryJobRecord{}, err
	}
	return job, nil
}

func (s *Server) listJobs(w http.ResponseWriter, r *http.Request, projectID string) {
	jobs, err := s.readQueryJobRecords(projectID)
	if err != nil {
		writeError(w, http.StatusInternalServerError, "backendError", "internal error")
		return
	}
	offset, err := rowOffsetFromRequest(r)
	if err != nil {
		writeError(w, http.StatusBadRequest, "invalid", err.Error())
		return
	}
	maxResults, err := maxResultsFromRequest(r)
	if err != nil {
		writeError(w, http.StatusBadRequest, "invalid", err.Error())
		return
	}
	if offset > len(jobs) {
		offset = len(jobs)
	}
	end := offset + maxResults
	if end > len(jobs) {
		end = len(jobs)
	}
	items := make([]jobResource, 0, end-offset)
	for _, job := range jobs[offset:end] {
		items = append(items, job.Job)
	}
	response := jobsListResponse{
		Kind: "bigquery#jobList",
		Jobs: items,
	}
	if end < len(jobs) {
		response.NextPageToken = strconv.Itoa(end)
	}
	writeJSON(w, http.StatusOK, response)
}

func (s *Server) cancelJob(w http.ResponseWriter, r *http.Request, projectID string, jobID string) {
	if err := validateResourceID(jobID, "job"); err != nil {
		writeError(w, http.StatusBadRequest, "invalid", err.Error())
		return
	}
	job, found, err := s.readQueryJob(projectID, jobID)
	if err != nil {
		writeError(w, http.StatusInternalServerError, "backendError", "internal error")
		return
	}
	if !found {
		writeError(w, http.StatusNotFound, "notFound", fmt.Sprintf("Not found: Job %s:%s", projectID, jobID))
		return
	}
	writeJSON(w, http.StatusOK, jobCancelResponse{
		Kind: "bigquery#jobCancelResponse",
		Job:  job.Job,
	})
}

func (s *Server) deleteJobMetadata(w http.ResponseWriter, r *http.Request, projectID string, jobID string) {
	if err := validateResourceID(jobID, "job"); err != nil {
		writeError(w, http.StatusBadRequest, "invalid", err.Error())
		return
	}
	path := s.queryJobPath(projectID, jobID)
	if err := os.Remove(path); errors.Is(err, os.ErrNotExist) {
		writeError(w, http.StatusNotFound, "notFound", fmt.Sprintf("Not found: Job %s:%s", projectID, jobID))
		return
	} else if err != nil {
		writeError(w, http.StatusInternalServerError, "backendError", "internal error")
		return
	}
	w.WriteHeader(http.StatusNoContent)
}

func (s *Server) getQueryResults(w http.ResponseWriter, r *http.Request, projectID string, jobID string) {
	if err := validateResourceID(jobID, "job"); err != nil {
		writeError(w, http.StatusBadRequest, "invalid", err.Error())
		return
	}
	job, found, err := s.readQueryJob(projectID, jobID)
	if err != nil {
		writeError(w, http.StatusInternalServerError, "backendError", "internal error")
		return
	}
	if !found {
		writeError(w, http.StatusNotFound, "notFound", fmt.Sprintf("Not found: Job %s:%s", projectID, jobID))
		return
	}
	offset, err := rowOffsetFromRequest(r)
	if err != nil {
		writeError(w, http.StatusBadRequest, "invalid", err.Error())
		return
	}
	maxResults, err := maxResultsFromRequest(r)
	if err != nil {
		writeError(w, http.StatusBadRequest, "invalid", err.Error())
		return
	}
	writeJSON(w, http.StatusOK, s.pageQueryResponse(job.Response, offset, maxResults))
}

func (s *Server) pageQueryResponse(response queryResponse, offset int, maxResults int) queryResponse {
	if offset < 0 {
		offset = 0
	}
	if maxResults <= 0 || maxResults > s.maxResultRows() {
		maxResults = s.maxResultRows()
	}
	if offset > len(response.Rows) {
		offset = len(response.Rows)
	}
	end := offset + maxResults
	if end > len(response.Rows) {
		end = len(response.Rows)
	}
	paged := response
	paged.Rows = response.Rows[offset:end]
	if end < len(response.Rows) {
		paged.PageToken = strconv.Itoa(end)
	} else {
		paged.PageToken = ""
	}
	return paged
}

File: services/bigquery/src/job_handlers.rs
//! Job handlers — port of `internal/services/bigquery/job_handlers.rs` plus
//! the inline job GET from legacy `handleJobs` (routes.rs, 3-part path).
//!
//! `insertJob` dispatches on the populated configuration exactly like legacy;
//! the copy/load/extract arms execute through `job_load_extract`.

use crate::model::{JobCancelResponse, JobInsertRequest, JobResource, JobsListResponse};
use crate::responses::ApiResponse;
use crate::server::Server;
use crate::validation::{
    decode_body, max_results_from_request, row_offset_from_request, validate_resource_id, Query,
};

impl Server {
    /// `POST /bigquery/v2/projects/{p}/jobs` (legacy `insertJob`).
    pub fn insert_job(&self, project_id: &str, body: &[u8]) -> ApiResponse {
        let request: JobInsertRequest = match decode_body(body, self.max_request_bytes()) {
            Ok(request) => request,
            Err(()) => return ApiResponse::error(400, "invalid", "invalid json request"),
        };
        if !request.job_reference.project_id.is_empty()
            && request.job_reference.project_id != project_id
        {
            return ApiResponse::error(
                400,
                "invalid",
                "jobReference.projectId must match request project",
            );
        }
        let config = request.configuration;
        if !config.query.query.trim().is_empty() {
            let use_legacy_sql = self.effective_use_legacy_sql(config.query.use_legacy_sql);
            return match self.create_query_job(
                project_id,
                &request.job_reference,
                config.query,
                0,
                true,
                config.dry_run,
                use_legacy_sql,
            ) {
                Err(message) => ApiResponse::error(400, "invalidQuery", &message),
                Ok(job) => {
                    self.emit_job_inserted(project_id, "query");
                    ApiResponse::json(200, &job.job)
                }
            };
        }
        if !config.copy.destination_table.table_id.is_empty() {
            return match self.create_copy_job(project_id, &request.job_reference, config.copy) {
                Err(message) => ApiResponse::error(400, "invalid", &message),
                Ok(job) => {
                    self.emit_job_inserted(project_id, "copy");
                    ApiResponse::json(200, &job.job)
                }
            };
        }
        if !config.load.destination_table.table_id.is_empty() {
            return match self.create_load_job(project_id, &request.job_reference, config.load) {
                Err(message) => ApiResponse::error(400, "invalid", &message),
                Ok(job) => {
                    self.emit_job_inserted(project_id, "load");
                    ApiResponse::json(200, &job.job)
                }
            };
        }
        if !config.extract.source_table.table_id.is_empty() {
            return match self.create_extract_job(project_id, &request.job_reference, config.extract)
            {
                Err(message) => ApiResponse::error(400, "invalid", &message),
                Ok(job) => {
                    self.emit_job_inserted(project_id, "extract");
                    ApiResponse::json(200, &job.job)
                }
            };
        }
        ApiResponse::error(
            400,
            "invalid",
            "configuration.query.query, configuration.copy, configuration.load, or configuration.extract is required",
        )
    }

    /// legacy `events.Emit(... "bigquery.job.inserted" ...)` in `insertJob`.
    fn emit_job_inserted(&self, project_id: &str, job_type: &str) {
        self.emit_event(
            "bigquery.job.inserted",
            serde_json::json!({"project": project_id, "jobType": job_type}),
        );
    }

    /// `GET /bigquery/v2/projects/{p}/jobs/{j}` — the inline handler in legacy
    /// `handleJobs` (routes.rs). Note: legacy does not `validateResourceID` the
    /// job id on this path; parity kept.
    pub fn get_job(&self, project_id: &str, job_id: &str) -> ApiResponse {
        match self.read_query_job(project_id, job_id) {
            Err(_) => ApiResponse::error(500, "backendError", "internal error"),
            Ok(None) => ApiResponse::error(
                404,
                "notFound",
                &format!("Not found: Job {project_id}:{job_id}"),
            ),
            Ok(Some(job)) => ApiResponse::json(200, &job.job),
        }
    }

    /// `GET /bigquery/v2/projects/{p}/jobs` (legacy `listJobs`).
    pub fn list_jobs(&self, project_id: &str, query: &Query) -> ApiResponse {
        let jobs = match self.read_query_job_records(project_id) {
            Ok(jobs) => jobs,
            Err(_) => return ApiResponse::error(500, "backendError", "internal error"),
        };
        let offset = match row_offset_from_request(query) {
            Ok(offset) => offset,
            Err(message) => return ApiResponse::error(400, "invalid", &message),
        };
        let max_results = match max_results_from_request(query) {
            Ok(max_results) => max_results,
            Err(message) => return ApiResponse::error(400, "invalid", &message),
        };
        let total = jobs.len() as i64;
        let offset = offset.min(total);
        let end = (offset + max_results).min(total);
        let items: Vec<JobResource> = jobs[offset as usize..end as usize]
            .iter()
            .map(|job| job.job.clone())
            .collect();
        let response = JobsListResponse {
            kind: "bigquery#jobList".to_string(),
            jobs: items,
            next_page_token: if end < total {
                end.to_string()
            } else {
                String::new()
            },
        };
        ApiResponse::json(200, &response)
    }

    /// `POST /bigquery/v2/projects/{p}/jobs/{j}/cancel` (legacy `cancelJob`).
    /// Every devcloud job is already DONE, so cancel just echoes it.
    pub fn cancel_job(&self, project_id: &str, job_id: &str) -> ApiResponse {
        if let Err(message) = validate_resource_id(job_id, "job") {
            return ApiResponse::error(400, "invalid", &message);
        }
        match self.read_query_job(project_id, job_id) {
            Err(_) => ApiResponse::error(500, "backendError", "internal error"),
            Ok(None) => ApiResponse::error(
                404,
                "notFound",
                &format!("Not found: Job {project_id}:{job_id}"),
            ),
            Ok(Some(job)) => ApiResponse::json(
                200,
                &JobCancelResponse {
                    kind: "bigquery#jobCancelResponse".to_string(),
                    job: job.job,
                },
            ),
        }
    }

    /// `DELETE /bigquery/v2/projects/{p}/jobs/{j}[/delete]`
    /// (legacy `deleteJobMetadata`).
    pub fn delete_job_metadata(&self, project_id: &str, job_id: &str) -> ApiResponse {
        if let Err(message) = validate_resource_id(job_id, "job") {
            return ApiResponse::error(400, "invalid", &message);
        }
        let path = self.query_job_path(project_id, job_id);
        match std::fs::remove_file(&path) {
            Err(err) if err.kind() == std::io::ErrorKind::NotFound => ApiResponse::error(
                404,
                "notFound",
                &format!("Not found: Job {project_id}:{job_id}"),
            ),
            Err(_) => ApiResponse::error(500, "backendError", "internal error"),
            Ok(()) => ApiResponse::no_content(),
        }
    }

    /// `GET /bigquery/v2/projects/{p}/queries/{j}` and
    /// `GET /bigquery/v2/projects/{p}/jobs/{j}/getQueryResults`
    /// (legacy `getQueryResults`, mounted on both paths).
    pub fn get_query_results(&self, project_id: &str, job_id: &str, query: &Query) -> ApiResponse {
        if let Err(message) = validate_resource_id(job_id, "job") {
            return ApiResponse::error(400, "invalid", &message);
        }
        let job = match self.read_query_job(project_id, job_id) {
            Err(_) => return ApiResponse::error(500, "backendError", "internal error"),
            Ok(None) => {
                return ApiResponse::error(
                    404,
                    "notFound",
                    &format!("Not found: Job {project_id}:{job_id}"),
                )
            }
            Ok(Some(job)) => job,
        };
        let offset = match row_offset_from_request(query) {
            Ok(offset) => offset,
            Err(message) => return ApiResponse::error(400, "invalid", &message),
        };
        let max_results = match max_results_from_request(query) {
            Ok(max_results) => max_results,
            Err(message) => return ApiResponse::error(400, "invalid", &message),
        };
        ApiResponse::json(
            200,
            &self.page_query_response(job.response, offset, max_results),
        )
    }
}
